### PR TITLE
Fix SpEL template evaluation and support external bean fallback

### LINKED ISSUE #1848: SpringBoot 3 Declarative HTTP client interface, incompatibility with SpringBoot Resilience4J annotations
Resilience4j version: latest

Java version: 17

Hello Resiliene4J team,

I am a happy user of SpringBoot Resilience4J, especially circuit breaker, fallback and annotation.

For instance, my typical use looks something like this:

```
    @CircuitBreaker(name = "myCircuitBreaker", fallbackMethod = "fallback")
    public Mono<SomeResponse> sendPostToSomeAPI(SomeRequest someRequest) {
        return webClient.mutate().baseUrl("http://some-url").build().post().uri("/api").body(BodyInserters.fromValue(someRequest)).retrieve().bodyToMono(SomeResponse.class);
    }

public Mono<Object> fallBack(Exception exception) {
        //complexe failure handling
        return Mono.error(new ...);
    }
```

Which is working very well, very happy, I see the circuit breaker, most of all, I do see the fallback methods beings invoked appropriately.

Recently, with Spring 3.0, there is this new feature of declarative http client, very well demonstrated here: https://www.youtube.com/watch?v=3NcmlrumSOc

With this new approach, I managed to refactor my old WebFlux WebClient into this new nicer way of doing things:

```
    @PostExchange(url = "http://some-url/api")
    Mono<SomeResponse> sendPostToSomeAPI(@RequestBody SomeRequest someRequest);
```

Unfortunately, now it is an interface, I am having trouble with defining the fallback method:

```
    @PostExchange(url = "http://some-url/api")
    @CircuitBreaker(name = "myCircuitBreaker", fallbackMethod = "fallback") //this is not working, since the real implementation of the fallback method is not, and cannot be in this new declarative interface
    Mono<SomeResponse> sendPostToSomeAPI(@RequestBody SomeRequest someRequest);
```

Hence, I would like to raise this issue, stating a possible incompatibility between the fallback annotation and its "in the same class fallback method" and the new Spring 3.0 declarative http clients.

Thank you

### LINKED ISSUE #2355: The SpEL expressions which not started from `#` doesn't evaluated
So if I define 
```java
    @CircuitBreaker(name = "#root.args[0] + '.one'", fallbackMethod = "#{'recovery'}")
    public String spelSync(String backend) {
        return syncError();
    }
```

it works fine and resolves to `backend.one`, but if I change it to 
```java
    @CircuitBreaker(name = "'one.' + #root.args[0]", fallbackMethod = "#{'recovery'}")
    public String spelSync(String backend) {
        return syncError();
    }
```
The expression has not been evaluated at all the name will be `'one.' + #root.args[0]`

### PR DESCRIPTION
**Pull request by LivingLikeKrillin**
## Summary

- **Fix SpEL `#{...}` template expressions** (#2355): Split the regex that previously treated both `${...}` and `#{...}` the same way. `#{...}` is now evaluated as a SpEL template with full method context (`#root.args`, `#root.methodName`, bean references), so expressions like `#{'one.' + #root.args[0]}` work correctly.
- **Support external bean fallback via `beanName::methodName` syntax** (#1848): When using interface-based declarative HTTP clients (Spring Boot 3+), fallback methods can now reference a method on a separate Spring bean — e.g., `@CircuitBreaker(name = "cb", fallbackMethod = "myFallbackBean::recover")`.
- **Properly unwrap AOP proxies** for external fallback beans using `AopProxyUtils.getSingletonTarget()`, so `FallbackMethod.getTarget()` correctly distinguishes the original target from the proxy.
- **Validate `beanName::methodName` format**: Empty method names (e.g., `"myBean::"`) are rejected with a clear error message instead of silently failing.
- **Document the new syntax** in the `fallbackMethod` Javadoc of all 6 annotations (CircuitBreaker, Retry, Bulkhead, TimeLimiter, RateLimiter, Timer).

Closes #1848
Closes #2355

## Test plan

- [x] Unit tests for `beanName::methodName` happy path (bean found, method exists)
- [x] Unit tests for `beanName::methodName` when method does not exist on the external bean
- [x] Unit tests for empty method name edge case (`"myBean::"`)
- [x] Unit tests for SpEL template expressions with `#root.args` context
- [x] Unit tests for SpEL template expressions with `@bean` references
- [ ] Existing tests remain green (spring6 module has pre-existing Mockito compatibility failures unrelated to this PR)

### CODE REVIEW

**Comment by LivingLikeKrillin** [comment]
Thanks for the review! I've addressed both findings in cc26021d:

1. **Multiple `::` separators** — Added a second-separator check that throws a clear `NoSuchMethodException` when the resolved value contains more than one `::`. Covered by `testPrimaryMethodExecutionWithMultipleSeparators` in both modules.
2. **Log message placeholder** — Lifted `beanName` to method scope and now log the extracted bean name instead of the raw `fallbackMethodValue`, so the message matches its placeholder.

Also added an explicit guard + test for the case where `beanName::methodName` is used but no `BeanFactory` was injected, so users get a clear error instead of silent fallthrough.

File: resilience4j-spring6/src/main/java/io/github/resilience4j/spring6/fallback/FallbackExecutor.java [hunk 1]
```diff
@@ -5,32 +5,66 @@
...
         FallbackMethod fallbackMethod = null;
         if (StringUtils.hasLength(fallbackMethodName)) {
             try {
+                Object original;
+                Object proxy;
+                int separatorIdx = fallbackMethodName.indexOf(BEAN_METHOD_SEPARATOR);
+                if (separatorIdx == 0) {
+                    throw new NoSuchMethodException(
+                        "Invalid fallbackMethod format: bean name is empty in '" + fallbackMethodValue + "'");
+                }
```
[the remark this answers is not in the document]
Comment (LivingLikeKrillin) [reply resilience4j-spring6/src/main/java/io/github/resilience4j/spring6/fallback/FallbackExecutor.java:48]: Added a second-separator lookup after the first `::` and throw `NoSuchMethodException` with a clear message. Covered by `testPrimaryMethodExecutionWithMultipleSeparators`.

File: resilience4j-spring6/src/main/java/io/github/resilience4j/spring6/fallback/FallbackExecutor.java [hunk 2]
```diff
@@ -5,32 +5,66 @@
...
+                    proxy = proceedingJoinPoint.getThis();
+                }
                 fallbackMethod = FallbackMethod
-                    .create(fallbackMethodName, method, proceedingJoinPoint.getArgs(), proceedingJoinPoint.getTarget(), proceedingJoinPoint.getThis());
+                    .create(fallbackMethodName, method, proceedingJoinPoint.getArgs(), original, proxy);
             } catch (NoSuchMethodException ex) {
                 logger.warn("No fallback method match found", ex);
+            } catch (org.springframework.beans.BeansException ex) {
+                logger.warn("Failed to resolve fallback bean '{}'", fallbackMethodValue, ex);
             }
```
[the remark this answers is not in the document]
Comment (LivingLikeKrillin) [reply resilience4j-spring6/src/main/java/io/github/resilience4j/spring6/fallback/FallbackExecutor.java]: Lifted `beanName` to method scope and the warn log now uses the extracted bean name instead of the raw `fallbackMethodValue`.

File: resilience4j-spring/src/main/java/io/github/resilience4j/fallback/FallbackExecutor.java [hunk 3]
```diff
@@ -5,32 +5,66 @@
...
         if (StringUtils.hasLength(fallbackMethodName)) {
             try {
+                Object original;
+                Object proxy;
+                int separatorIdx = fallbackMethodName.indexOf(BEAN_METHOD_SEPARATOR);
+                if (separatorIdx == 0) {
+                    throw new NoSuchMethodException(
+                        "Invalid fallbackMethod format: bean name is empty in '" + fallbackMethodValue + "'");
+                }
+                if (separatorIdx > 0 && beanFactory != null) {
```
[the remark this answers is not in the document]
Comment (LivingLikeKrillin) [reply resilience4j-spring/src/main/java/io/github/resilience4j/fallback/FallbackExecutor.java]: Added a second-separator lookup after the first `::` and throw `NoSuchMethodException` with a clear message. Covered by `testPrimaryMethodExecutionWithMultipleSeparators`.

File: resilience4j-spring/src/main/java/io/github/resilience4j/fallback/FallbackExecutor.java [hunk 4]
```diff
@@ -5,32 +5,66 @@
...
+                    proxy = proceedingJoinPoint.getThis();
+                }
                 fallbackMethod = FallbackMethod
-                    .create(fallbackMethodName, method, proceedingJoinPoint.getArgs(), proceedingJoinPoint.getTarget(), proceedingJoinPoint.getThis());
+                    .create(fallbackMethodName, method, proceedingJoinPoint.getArgs(), original, proxy);
             } catch (NoSuchMethodException ex) {
                 logger.warn("No fallback method match found", ex);
+            } catch (org.springframework.beans.BeansException ex) {
+                logger.warn("Failed to resolve fallback bean '{}'", fallbackMethodValue, ex);
             }
```
[the remark this answers is not in the document]
Comment (LivingLikeKrillin) [reply resilience4j-spring/src/main/java/io/github/resilience4j/fallback/FallbackExecutor.java]: Lifted `beanName` to method scope and the warn log now uses the extracted bean name instead of the raw `fallbackMethodValue`.

**Comment by LivingLikeKrillin** [comment]
Hi @RobWin, I've addressed the Copilot review feedback in cc26021d (multiple-separator validation + clearer log message on bean resolution failure). The PR is ready for another look whenever you have time. Thanks!

File: resilience4j-spring6/src/main/java/io/github/resilience4j/spring6/fallback/FallbackExecutor.java [hunk 5]
```diff
@@ -5,32 +5,79 @@
...
+                    fallbackMethodName = fallbackMethodName.substring(separatorIdx + BEAN_METHOD_SEPARATOR.length());
+                    if (!StringUtils.hasLength(fallbackMethodName)) {
+                        throw new NoSuchMethodException(
+                            "Invalid fallbackMethod format: expected 'beanName::methodName' but got '" + fallbackMethodValue + "'");
+                    }
+                    Object fallbackBean = beanFactory.getBean(beanName);
+                    proxy = fallbackBean;
+                    Object singletonTarget = AopProxyUtils.getSingletonTarget(fallbackBean);
+                    original = (singletonTarget != null) ? singletonTarget : fallbackBean;
+                } else {
```
[the remark this answers is not in the document]
Comment (LivingLikeKrillin) [reply resilience4j-spring6/src/main/java/io/github/resilience4j/spring6/fallback/FallbackExecutor.java:78]: Added testPrimaryMethodExecutionWithAopProxiedFallbackBean in 38d31b29.

File: resilience4j-spring6/src/test/java/io/github/resilience4j/spring6/fallback/FallbackExecutorTest.java [hunk 6]
```diff
@@ -106,4 +114,151 @@ public String getNameInvalidFallback() {
...
+    @Test
+    public void testPrimaryMethodExecutionWithFallbackBeanNotFound() throws Throwable {
+        Method method = this.getClass().getMethod("getName", String.class);
+        final CheckedSupplier<Object> primaryFunction = () -> getName("Name");
+        final String fallbackMethodValue = "myFallbackBean::nonExistentMethod";
+
+        when(proceedingJoinPoint.getArgs()).thenReturn(new Object[]{});
+        when(spelResolver.resolve(method, proceedingJoinPoint.getArgs(), fallbackMethodValue)).thenReturn(fallbackMethodValue);
+        when(beanFactory.getBean("myFallbackBean")).thenReturn(new ExternalFallbackBean());
+
```
[the remark this answers is not in the document]
Comment (LivingLikeKrillin) [reply resilience4j-spring6/src/test/java/io/github/resilience4j/spring6/fallback/FallbackExecutorTest.java:153]: Renamed to testPrimaryMethodExecutionWithFallbackMethodNotFound in 38d31b29.

File: resilience4j-spring/src/main/java/io/github/resilience4j/fallback/FallbackExecutor.java [hunk 7]
```diff
@@ -5,32 +5,79 @@
...
+                    fallbackMethodName = fallbackMethodName.substring(separatorIdx + BEAN_METHOD_SEPARATOR.length());
+                    if (!StringUtils.hasLength(fallbackMethodName)) {
+                        throw new NoSuchMethodException(
+                            "Invalid fallbackMethod format: expected 'beanName::methodName' but got '" + fallbackMethodValue + "'");
+                    }
+                    Object fallbackBean = beanFactory.getBean(beanName);
+                    proxy = fallbackBean;
+                    Object singletonTarget = AopProxyUtils.getSingletonTarget(fallbackBean);
+                    original = (singletonTarget != null) ? singletonTarget : fallbackBean;
+                } else {
```
[the remark this answers is not in the document]
Comment (LivingLikeKrillin) [reply resilience4j-spring/src/main/java/io/github/resilience4j/fallback/FallbackExecutor.java:78]: Added testPrimaryMethodExecutionWithAopProxiedFallbackBean in 38d31b29.

File: resilience4j-spring/src/test/java/io/github/resilience4j/fallback/FallbackExecutorTest.java [hunk 8]
```diff
@@ -104,4 +112,151 @@ public String getNameInvalidFallback() {
...
+    @Test
+    public void testPrimaryMethodExecutionWithFallbackBeanNotFound() throws Throwable {
+        Method method = this.getClass().getMethod("getName", String.class);
+        final CheckedSupplier<Object> primaryFunction = () -> getName("Name");
+        final String fallbackMethodValue = "myFallbackBean::nonExistentMethod";
+
+        when(proceedingJoinPoint.getArgs()).thenReturn(new Object[]{});
+        when(spelResolver.resolve(method, proceedingJoinPoint.getArgs(), fallbackMethodValue)).thenReturn(fallbackMethodValue);
+        when(beanFactory.getBean("myFallbackBean")).thenReturn(new ExternalFallbackBean());
+
```
[the remark this answers is not in the document]
Comment (LivingLikeKrillin) [reply resilience4j-spring/src/test/java/io/github/resilience4j/fallback/FallbackExecutorTest.java:151]: Renamed to testPrimaryMethodExecutionWithFallbackMethodNotFound in 38d31b29.

**Comment by LivingLikeKrillin** [comment]
Hi @RobWin, addressed the Copilot review feedback in 38d31b29 (multiple-separator validation, AOP proxy test coverage, ArgumentCaptor verification, narrowed exception handling, BeanFactory ctor overload). Ready for another look when you have time. Thanks!
